### PR TITLE
Implement table<T> in semtypes

### LINKED ISSUE #713: Implement `table<T>` in semtypes
This issue is to implement just `table<T>` in semtypes, and in the front-end as a type descriptor.

This is useful, because it allows us to properly implement `anydata` which includes `table<map<anydata>>`, which in turn allows us to implement `record { }` #573 

Currently we approximate `anydata` in the compiler (for the purposes of checking when `==` is allowed), by including all of the table basic type, but this isn't a good enough approximation for semtype.

Following on from this would be:
* semtypes key-constraint: `table<T> key<K>` or `key(x, y, z)`
* full support for table values in the compiler

### PR DESCRIPTION
**Pull request by ushirask**
Resolve #713 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/parseTypeDesc.bal [hunk 1]
```diff
@@ -174,6 +174,12 @@ function parsePrimaryTypeDesc(Tokenizer tok) returns TypeDesc|err:Syntax {
         "record" => {
             return parseRecordTypeDesc(tok, startPos);
         }
+        "table" => {
+            check tok.advance();
+            var row = check parseTypeParam(tok);
```
Comment (rdhananjaya) [inline compiler/modules/front.syntax/parseTypeDesc.bal]: I see that the `var` is used in other places in this file, shall we use to type-desc here.

File: compiler/modules/front.syntax/parseTypeDesc.bal [hunk 2]
```diff
@@ -174,6 +174,12 @@ function parsePrimaryTypeDesc(Tokenizer tok) returns TypeDesc|err:Syntax {
...
+            Position endPos = tok.previousEndPos();
+            return { startPos, endPos, row};
```
Comment (rdhananjaya) [inline compiler/modules/front.syntax/parseTypeDesc.bal]: Missing space before closing brace.

File: compiler/modules/types/table.bal [hunk 3]
```diff
@@ -0,0 +1,61 @@
+public function tableDef(SemType constituentType) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/table.bal]: We need a better name for this function.  Let's call it `tableContaining` (but if you have a better idea, please propose it).

Use `memberType` not `constituentType` (the latter is only for sequences).

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/types/table.bal [hunk 4]
```diff
@@ -0,0 +1,34 @@
+public function tableContaining(SemType memberType) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/table.bal:8]: Put a comment
```
// Implementation specific to basic type table.
```

File: compiler/testSuite/semtype/table-t.bal [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+|};
+
+type READ readonly;
+
+// @type W < T
+// @type W < READ
+// @type T < READ
+// @type W < Y
+// @type Y <> T
+type T table<R> & readonly | table<R1> & readonly;
```
Comment (jclark) [inline compiler/testSuite/semtype/table-t.bal:21]: We need a test case where there is a subtype relationship between the records (e.g. same fields but one member type is a subtype of the other)

File: compiler/modules/front.syntax/ast.bal [hunk 6]
```diff
@@ -537,6 +537,12 @@ public type TypeDescRef record {|
     Position pos;
 |};
 
+public type TableTypeDesc record {|
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: Put this before TypeDescRef.

File: compiler/modules/front.syntax/scan.bal [hunk 7]
```diff
@@ -133,7 +133,8 @@ final readonly & Keyword[] keywords = [
     "type",
     "typedesc",
     "while",
-    "xml"
+    "xml",
+    "table"
```
Comment (jclark) [inline compiler/modules/front.syntax/scan.bal]: Do you notice anything about the other values in this list?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/scan.bal (hunk 7 above)
Comment (ushirask) [reply compiler/modules/front.syntax/scan.bal]: should they be put in alphabetical order? I'll change it, thanks

File: compiler/modules/types/table.bal [hunk 8]
```diff
@@ -0,0 +1,36 @@
+// Implementation specific to basic type table.
+
+public function tableContaining(SemType memberType) returns SemType {
+    SubtypeData ro = subtypeData(memberType, UT_MAPPING_RO);
+    SubtypeData rw = subtypeData(memberType, UT_MAPPING_RW);
```
Comment (jclark) [inline compiler/modules/types/table.bal]: Unfortunately, it's not this simple https://github.com/ballerina-platform/nballerina/issues/713#issuecomment-985316349

File: compiler/modules/types/table.bal [hunk 9]
```diff
@@ -0,0 +1,71 @@
+// Implementation specific to basic type table.
+
+public type RwTableSubtype record {|
+    Bdd ro;
+    Bdd rw;
+|};
+
+public function tableContaining(SemType memberType) returns SemType {
+    Bdd ro = <Bdd> subtypeData(memberType, UT_MAPPING_RO);
```
Comment (jclark) [inline compiler/modules/types/table.bal]: I don't put a space after `>`.

File: compiler/modules/types/table.bal [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    Bdd rw = <Bdd> subtypeData(memberType, UT_MAPPING_RW);
+    SubtypeData rwt = createRwTableSubtype(ro, rw);
+    return createComplexSemType(0, [[UT_TABLE_RO, ro], [UT_TABLE_RW, rwt]]);
+}
+
+function rwTableSubtypeUnion(SubtypeData t1, SubtypeData t2) returns SubtypeData {
+    RwTableSubtype rwt1 = <RwTableSubtype> t1;
+    RwTableSubtype rwt2 = <RwTableSubtype> t2;
+    Bdd ro = bddUnion(rwt1.ro, rwt2.ro);
```
Comment (jclark) [inline compiler/modules/types/table.bal]: Inline ro and rw here and following.

File: compiler/modules/types/table.bal [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+function rwTableSubtypeEmpty(Context cx, SubtypeData t) returns boolean {
+    RwTableSubtype rwt = <RwTableSubtype> t;
+    boolean ro = mappingRoSubtypeIsEmpty(cx, rwt.ro);
+    boolean rw = mappingSubtypeIsEmpty(cx, rwt.rw);
+    return ro && rw;
+}
+
+function createRwTableSubtype(Bdd ro, Bdd rw) returns SubtypeData {
+    return {ro, rw};
```
Comment (jclark) [inline compiler/modules/types/table.bal]: If ro and rw are false, then this is false. Likewise for true. You need to use this with union/intersect/diff/complement.

File: compiler/modules/types/table.bal [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+public type RwTableSubtype record {|
+    Bdd ro;
+    Bdd rw;
+|};
+
+public function tableContaining(SemType memberType) returns SemType {
+    Bdd ro = <Bdd> subtypeData(memberType, UT_MAPPING_RO);
+    Bdd rw = <Bdd> subtypeData(memberType, UT_MAPPING_RW);
+    SubtypeData rwt = createRwTableSubtype(ro, rw);
+    return createComplexSemType(0, [[UT_TABLE_RO, ro], [UT_TABLE_RW, rwt]]);
```
Comment (jclark) [inline compiler/modules/types/table.bal]: You have to handle the cases where `ro` and `rwt` are true or false. In these cases, the ComplexSemType does not have a SubtypeData member for them (just bits in all).

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/types/table.bal [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+function rwTableSubtypeEmpty(Context cx, SubtypeData t) returns boolean {
+    RwTableSubtype rwt = <RwTableSubtype>t;
+    return mappingRoSubtypeIsEmpty(cx, rwt.ro) && mappingSubtypeIsEmpty(cx, rwt.rw);
+}
+
+function createRwTableSubtype(Bdd ro, Bdd rw) returns SubtypeData {
+    if ro == false && rw == false {
+        return false;
+    } else if ro == true && rw == true {
```
Comment (jclark) [inline compiler/modules/types/table.bal]: `else` on newline

File: compiler/modules/types/table.bal [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        return true;
+    }
+    return {ro, rw};
```
Comment (jclark) [inline compiler/modules/types/table.bal]: spaces inside braces

File: compiler/modules/types/table.bal [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    return createRwTableSubtype(bddComplement(rwt.ro), bddComplement(rwt.rw));
+}
+
+function rwTableSubtypeEmpty(Context cx, SubtypeData t) returns boolean {
+    RwTableSubtype rwt = <RwTableSubtype>t;
+    return mappingRoSubtypeIsEmpty(cx, rwt.ro) && mappingSubtypeIsEmpty(cx, rwt.rw);
+}
+
+function createRwTableSubtype(Bdd ro, Bdd rw) returns SubtypeData {
+    if ro == false && rw == false {
```
Comment (jclark) [inline compiler/modules/types/table.bal]: We can do this two tests more neatly as:

```
if ro is boolean && ro == rw {
    return ro;
}
```

File: compiler/modules/front/resolveTypes.bal [hunk 16]
```diff
@@ -278,6 +278,16 @@ function resolveTypeDesc(ModuleSymbols mod, s:ModuleLevelDefn modDefn, int depth
         }
         return t:xmlSequence(t);
     }
+    if td is s:TableTypeDesc {
+        t:SemType t = check resolveTypeDesc(mod, modDefn, depth, td.row);
+        d:Location loc =  d:location(modDefn.part.file, td.startPos, td.endPos);
```
Comment (jclark) [inline compiler/modules/front/resolveTypes.bal]: Put the declaration of `loc` inside the `if`, and do the same for the XML case above.

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**